spec/requests/topics_controller_spec.rb

```diff
@@ -4,158 +4,1464 @@
+            expect(response.status).to eq(value)
+          end
+        end
+      end
+
+      context 'anonymous' do
+        expected = {
+          normal_topic: 200,
+          secure_topic: 403,
+          private_topic: 403,
+          deleted_topic: 410,
+          deleted_secure_topic: 403,
+          deleted_private_topic: 403,
+          nonexist: 404
+        }
+        include_examples "various scenarios", expected
+      end
+
+      context 'anonymous with login required' do
+        before do
+          SiteSetting.login_required = true
+        end
+        expected = {
+          normal_topic: 302,
+          secure_topic: 302,
+          private_topic: 302,
+          deleted_topic: 302,
+          deleted_secure_topic: 302,
+          deleted_private_topic: 302,
+          nonexist: 302
+        }
+        include_examples "various scenarios", expected
+      end
+
+      context 'normal user' do
+        before do
+          sign_in(Fabricate(:user))
+        end
+
+        expected = {
+          normal_topic: 200,
+          secure_topic: 403,
+          private_topic: 403,
+          deleted_topic: 410,
+          deleted_secure_topic: 403,
+          deleted_private_topic: 403,
+          nonexist: 404
+        }
+        include_examples "various scenarios", expected
+      end
+
+      context 'allowed user' do
+        before do
+          sign_in(allowed_user)
+        end
+
+        expected = {
+          normal_topic: 200,
+          secure_topic: 200,
+          private_topic: 200,
+          deleted_topic: 410,
+          deleted_secure_topic: 410,
+          deleted_private_topic: 410,
+          nonexist: 404
+        }
+        include_examples "various scenarios", expected
+      end
+
+      context 'moderator' do
+        before do
+          sign_in(Fabricate(:moderator))
+        end
+
+        expected = {
+          normal_topic: 200,
+          secure_topic: 403,
+          private_topic: 403,
+          deleted_topic: 200,
+          deleted_secure_topic: 403,
+          deleted_private_topic: 403,
+          nonexist: 404
+        }
+        include_examples "various scenarios", expected
+      end
+
+      context 'admin' do
+        before do
+          sign_in(Fabricate(:admin))
+        end
+
+        expected = {
+          normal_topic: 200,
+          secure_topic: 200,
+          private_topic: 200,
+          deleted_topic: 200,
+          deleted_secure_topic: 200,
+          deleted_private_topic: 200,
+          nonexist: 404
+        }
+        include_examples "various scenarios", expected
+      end
+    end
+
+    it 'records a view' do
+      expect do
+        get "/t/#{topic.slug}/#{topic.id}.json"
+      end.to change(TopicViewItem, :count).by(1)
+    end
+
+    it 'records incoming links' do
+      user = Fabricate(:user)
+
+      get "/t/#{topic.slug}/#{topic.id}", params: {
+        u: user.username
+      }
+
+      expect(IncomingLink.count).to eq(1)
```

Comment:
We should use `expect {}.to_change {}.by(1)`. Otherwise if an incoming link already exists in the test db, this test becomes a false positive.